File: backend/app/schemas/station.py
"""Station overview and environmental Pydantic schemas."""

from typing import Optional
from pydantic import BaseModel, Field

from app.models.enums import EnvironmentMode, StationStatus
from app.schemas.common import ProvenanceSchema


class AmbientWeatherSchema(BaseModel):
    """Current station weather observation."""

    temperature_celsius: float
    wind_speed_knots: float
    wind_chill_celsius: float
    conditions: str
    provenance: ProvenanceSchema


class SubsystemSummaryItem(BaseModel):
    """High-level operational health indicator for a station subsystem."""

    code: str
    name: str
    status: str
    health_score: int = Field(ge=0, le=100)


class CriticalEventItem(BaseModel):
    """Urgent operational anomaly or active incident requiring operator attention."""

    id: str
    title: str
    severity: str
    status: str
    asset_id: Optional[str] = None
    location: Optional[str] = None
    description: str


class StationOverviewResponse(BaseModel):
    """High-level situation awareness payload for Station Command Center."""

    station_id: str
    name: str
    status: StationStatus
    environment_mode: EnvironmentMode
    overall_health_score: int = Field(ge=0, le=100)
    active_incidents_count: int
    fuel_runway_days: Optional[float] = None
    fuel_quantity_liters: Optional[float] = None
    connectivity_status: str = "ONLINE"
    ambient_weather: AmbientWeatherSchema
    subsystem_summary: list[SubsystemSummaryItem]
    critical_events: list[CriticalEventItem] = []


class OperationalCapabilityItem(BaseModel):
    """Derived operational capability headroom score and state for a station domain."""

    domain: str  # "ENERGY_RESILIENCE", "COMMS_CONTINUITY", "SCIENCE_CONTINUITY", "LIFE_SUPPORT", "RECOVERY_BUFFER"
    name: str
    headroom_score: int = Field(ge=0, le=100)
    status: str  # "NOMINAL", "CONSTRAINED", "CRITICAL"
    summary: str
    calculation_basis: Optional[str] = None
    metrics: dict[str, str | int | float | bool] = Field(default_factory=dict)


class OperationalDifferenceItem(BaseModel):
    """Structured operational divergence between compared stations."""

    dimension: str  # "STATION_HEALTH", "FUEL_RUNWAY", "ASSET_RISK", "COMMUNICATIONS", "CRITICAL_SPARES", "WEATHER_EXPOSURE"
    title: str
    station_a_value: str
    station_b_value: str
    delta_summary: str
    pressure_direction: str  # "BHARATI_HIGHER", "MAITRI_HIGHER", "BALANCED"
    significance: str  # "CRITICAL", "MODERATE", "INFORMATIONAL"


class CoordinationConstraintItem(BaseModel):
    """Explicit physical, meteorological, or telecommunications constraint."""

    constraint_type: str  # "LOGISTICS_DISTANCE", "WEATHER_FLIGHT_WINDOW", "COMMS_ASYMMETRY", "TRAVERSE_SEASON"
    name: str
    status: str  # "RESTRICTED", "IMPASSABLE", "NOMINAL", "DEGRADED"
    impact: str
    details: str
    provenance_type: str = "DOCUMENTED_GEOGRAPHY"  # "DOCUMENTED_GEOGRAPHY", "MODELED_OPERATIONAL_RULE", "MODELED_SYSTEM_PROFILE"
    validation_status: str = "VERIFIED_RESEARCH"  # "VERIFIED_RESEARCH", "REQUIRES_FUTURE_VALIDATION"


class CrossStationConsiderationItem(BaseModel):
    """Advisory operational consideration for human operator decision support (non-actuating)."""

    id: str
    category: str  # "RESOURCE_SUPPORT", "RECOVERY_ALIGNMENT", "COMMS_READINESS", "RISK_MITIGATION"
    title: str
    recommendation: str
    rationale: str
    prerequisites: list[str] = Field(default_factory=list)
    feasibility_status: str  # "FEASIBLE_WITH_CONSTRAINTS", "RESTRICTED", "ADVISORY_ONLY"


class RecoveryChainItem(BaseModel):
    """Structured end-to-end recovery chain distinguishing technical, material, and logistics constraints."""

    station_id: str
    asset_id: str
    asset_code: str = "G-02"
    asset_name: str
    technical_condition: str
    material_constraint: str
    local_availability: str = "0 units available (Stockout)"
    local_stock_quantity: int = 0
    maintenance_constraint: str = "BLOCKED (MWO-2026-089 awaiting parts)"
    maintenance_status: str = "BLOCKED"
    work_order_id: Optional[str] = None
    resupply_dependency: str
    candidate_support_station: Optional[str] = None
    recovery_status: str = "CONSTRAINED"  # "CONSTRAINED", "NOMINAL", "MONITORING"
    recovery_exposure: str = "Loss of N+1 redundancy; single-fault vulnerability"
    operational_exposure: str = "Single generator G-01 dependency; secondary heating loop exposed to freeze-out"
    timing_confidence: str = "Requires future validation"
    timing_disclaimer: str = "Recovery remains constrained until the required resource becomes available. Repair duration requires post-delivery mechanical inspection."


class StationPortfolioItem(BaseModel):
    """Summary of a station's canonical state within the portfolio comparison."""

    station_id: str
    code: str
    name: str
    status: StationStatus
    overall_health: int
    fuel_runway_days: Optional[float] = None
    fuel_quantity_liters: Optional[float] = None
    temperature_celsius: float
    wind_speed_knots: float
    conditions: str
    comms_status: str
    active_incidents_count: int
    critical_spares_available: int
    capabilities: list[OperationalCapabilityItem] = Field(default_factory=list)


class StationComparisonResponse(BaseModel):
    """Deterministic comparison between two research stations."""

    station_a: StationPortfolioItem
    station_b: StationPortfolioItem
    capabilities_summary: list[dict[str, str | int | float]] = Field(default_factory=list)
    differences: list[OperationalDifferenceItem] = Field(default_factory=list)
    constraints: list[CoordinationConstraintItem] = Field(default_factory=list)
    considerations: list[CrossStationConsiderationItem] = Field(default_factory=list)
    recovery_chain: list[RecoveryChainItem] = Field(default_factory=list)
    higher_pressure_station_id: str
    pressure_rationale: str
    provenance: ProvenanceSchema

File: frontend/src/lib/api/station.ts
/**
 * Station Overview, Events, and Multi-Station Coordination API
 * Owner: Kshitij (Product Owner & System Integration Lead)
 */

import { API_BASE, handleApiResponse, type Provenance, type StationStatus, type EnvironmentMode } from "./client";

export interface AmbientWeather {
  temperature_celsius: number;
  wind_speed_knots: number;
  wind_chill_celsius: number;
  conditions: string;
  provenance: Provenance;
}

export interface SubsystemSummaryItem {
  code: string;
  name: string;
  status: string;
  health_score: number;
}

export interface CriticalEventItem {
  id: string;
  title: string;
  severity: string;
  status: string;
  asset_id?: string | null;
  location?: string | null;
  description: string;
}

export interface StationOverview {
  station_id: string;
  name: string;
  status: StationStatus;
  environment_mode: EnvironmentMode;
  overall_health_score: number;
  active_incidents_count: number;
  fuel_runway_days?: number | null;
  fuel_quantity_liters?: number | null;
  connectivity_status: string;
  ambient_weather: AmbientWeather;
  subsystem_summary: SubsystemSummaryItem[];
  critical_events: CriticalEventItem[];
}

export interface OperationalCapabilityItem {
  domain: string;
  name: string;
  headroom_score: number;
  status: "NOMINAL" | "CONSTRAINED" | "CRITICAL" | string;
  summary: string;
  calculation_basis?: string;
  metrics: Record<string, string | number | boolean>;
}

export interface OperationalDifferenceItem {
  dimension: string;
  title: string;
  station_a_value: string;
  station_b_value: string;
  delta_summary: string;
  pressure_direction: "BHARATI_HIGHER" | "MAITRI_HIGHER" | "BALANCED" | string;
  significance: "CRITICAL" | "MODERATE" | "INFORMATIONAL" | string;
}

export interface CoordinationConstraintItem {
  constraint_type: string;
  name: string;
  status: "RESTRICTED" | "IMPASSABLE" | "NOMINAL" | "DEGRADED" | string;
  impact: string;
  details: string;
  provenance_type?: string;
  validation_status?: string;
}

export interface CrossStationConsiderationItem {
  id: string;
  category: string;
  title: string;
  recommendation: string;
  rationale: string;
  prerequisites: string[];
  feasibility_status: "FEASIBLE_WITH_CONSTRAINTS" | "RESTRICTED" | "ADVISORY_ONLY" | string;
}

export interface RecoveryChainItem {
  station_id: string;
  asset_id: string;
  asset_code: string;
  technical_condition: string;
  material_constraint: string;
  local_availability: string;
  maintenance_constraint: string;
  resupply_dependency: string;
  recovery_exposure: string;
  recovery_status: "CONSTRAINED" | "NOMINAL" | "BLOCKED" | "WATCH" | string;
  timing_confidence: string;
  timing_disclaimer: string;
}

export interface StationPortfolioItem {
  station_id: string;
  code: string;
  name: string;
  status: StationStatus;
  overall_health: number;
  fuel_runway_days?: number | null;
  fuel_quantity_liters?: number | null;
  temperature_celsius: number;
  wind_speed_knots: number;
  conditions: string;
  comms_status: string;
  active_incidents_count: number;
  critical_spares_available: number;
  capabilities: OperationalCapabilityItem[];
}

export interface StationComparisonResponse {
  station_a: StationPortfolioItem;
  station_b: StationPortfolioItem;
  capabilities_summary: Array<Record<string, string | number>>;
  differences: OperationalDifferenceItem[];
  constraints: CoordinationConstraintItem[];
  considerations: CrossStationConsiderationItem[];
  recovery_chain?: RecoveryChainItem[];
  higher_pressure_station_id: string;
  pressure_rationale: string;
  provenance: Provenance;
}

export interface OperationalEvent {
  id: string;
  timestamp: string;
  station_id: string;
  event_type: string;
  entity_type: string;
  entity_id: string;
  title: string;
  summary: string;
  severity: "INFO" | "WARNING" | "CRITICAL" | "SYSTEM";
  truth_type: string;
  metadata?: Record<string, unknown> | null;
  is_synthetic: boolean;
}

export interface EventListResponse {
  events: OperationalEvent[];
  total_count: number;
  station_id: string;
  simulation_active: boolean;
  last_updated: string;
}

export interface CausalStageItem {
  stage: "CHANGE" | "CONTEXT" | "DEPENDENCY" | "RISK" | "CONSEQUENCE" | "SCENARIO" | "ACTION" | string;
  title: string;
  headline: string;
  description: string;
  severity: "CRITICAL" | "WARNING" | "NOMINAL" | "INFO" | string;
  truth_type: "MEASURED" | "DERIVED" | "SCENARIO" | "FORECAST" | string;
  supporting_metrics: Record<string, any>;
  target_route?: string | null;
  action_label?: string | null;
}

export interface OperationalDecisionItem {
  id: string;
  title: string;
  rationale: string;
  action_type: "INSPECT" | "SIMULATE" | "RECOVERY" | "RESILIENCE" | "ADVISORY" | string;
  target_route: string;
  button_label: string;
  is_primary: boolean;
}

export interface OperationalInsightResponse {
  station_id: string;
  station_name: string;
  primary_condition_id: string;
  severity: "CRITICAL" | "WARNING" | "NOMINAL" | string;
  status_label: string;
  headline: string;
  summary: string;
  causal_chain: CausalStageItem[];
  decisions: OperationalDecisionItem[];
  provenance: Provenance;
}

export async function fetchStationOverview(stationId: string): Promise<StationOverview> {
  const res = await fetch(`${API_BASE}/station/overview?station_id=${encodeURIComponent(stationId)}`);
  return handleApiResponse<StationOverview>(res, `Failed to fetch station overview for ${stationId}`);
}

export async function fetchStationComparison(
  stationA: string = "STATION-BHARATI",
  stationB: string = "STATION-MAITRI"
): Promise<StationComparisonResponse> {
  const url = `${API_BASE}/station/comparison?station_a_id=${encodeURIComponent(stationA)}&station_b_id=${encodeURIComponent(stationB)}`;
  const res = await fetch(url);
  return handleApiResponse<StationComparisonResponse>(res, "Failed to fetch station comparison");
}

export async function evaluateStationComparison(
  stationA: string = "STATION-BHARATI",
  stationB: string = "STATION-MAITRI"
): Promise<StationComparisonResponse> {
  const url = `${API_BASE}/station/comparison/evaluate?station_a_id=${encodeURIComponent(stationA)}&station_b_id=${encodeURIComponent(stationB)}`;
  const res = await fetch(url, { method: "POST" });
  return handleApiResponse<StationComparisonResponse>(res, "Failed to evaluate station comparison");
}

export async function fetchOperationalEvents(params?: {
  station_id?: string;
  severity?: string;
  event_type?: string;
  limit?: number;
  offset?: number;
}): Promise<EventListResponse> {
  const station = params?.station_id ?? "STATION-BHARATI";
  let url = `${API_BASE}/events?station_id=${encodeURIComponent(station)}`;
  if (params?.severity) url += `&severity=${encodeURIComponent(params.severity)}`;
  if (params?.event_type) url += `&event_type=${encodeURIComponent(params.event_type)}`;
  if (params?.limit) url += `&limit=${params.limit}`;
  if (params?.offset) url += `&offset=${params.offset}`;
  const res = await fetch(url);
  return handleApiResponse<EventListResponse>(res, "Failed to fetch operational events");
}

export async function simulateDemoEvent(
  stepIndex?: number,
  stationId: string = "STATION-BHARATI"
): Promise<OperationalEvent> {
  const res = await fetch(`${API_BASE}/events/simulate`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ step_index: stepIndex, station_id: stationId }),
  });
  return handleApiResponse<OperationalEvent>(res, "Failed to simulate demo event");
}

export async function resetDemoEvents(
  stationId: string = "STATION-BHARATI"
): Promise<{ status: string; message: string; events_reset_count: number }> {
  const res = await fetch(`${API_BASE}/events/reset?station_id=${encodeURIComponent(stationId)}`, {
    method: "POST",
  });
  return handleApiResponse<{ status: string; message: string; events_reset_count: number }>(res, "Failed to reset events");
}

export async function fetchOperationalIntelligence(
  stationId: string = "STATION-BHARATI"
): Promise<OperationalInsightResponse> {
  const url = `${API_BASE}/intelligence/narrative?station_id=${encodeURIComponent(stationId)}`;
  const res = await fetch(url);
  return handleApiResponse<OperationalInsightResponse>(res, "Failed to fetch operational intelligence");
}
